src/lib/core/fiber.c

```diff
@@ -1607,6 +1618,8 @@ struct cord_thread_arg
 void *cord_thread_func(void *p)
 {
 	struct cord_thread_arg *ct_arg = (struct cord_thread_arg *) p;
+	if (is_shutting_down)
+		return NULL;
```

Comment:
- We wait for the memtx checkpoint thread to exit before proceeding to shutdown, see `memtx_engine_shutdown`. I don't understand what can go wrong. Could you elaborate?
- We include bug fixes into changelogs. Please add a changelog entry.
- Can you add a test for this bug fix?

Reply:
What I see with simple logging through the `tarantool_free()` - before and after appropriate calls:
```
[092] [Instance "replica_auth" killed by signal: 11 (SIGSEGV)]
[092] 
[092] Last 15 lines of Tarantool Log file [Instance "replica_auth"][/tmp/t/092_replication/replica_auth.log]:
[092] 2022-11-15 00:01:29.678 [73463] main/112/applier/cluster@unix/:/tmp/t/092_replication/master.i I> update replication_synchro_quorum = 1
[092] 2022-11-15 00:01:29.678 [73463] main/112/applier/cluster@unix/:/tmp/t/092_replication/master.i I> initial data received
[092] 2022-11-15 00:01:29.678 [73463] main/112/applier/cluster@unix/:/tmp/t/092_replication/master.i I> assigned id 2 to replica 1aa02019-8281-4817-b461-b4f11a0c3eae
[092] 2022-11-15 00:01:29.678 [73463] main/112/applier/cluster@unix/:/tmp/t/092_replication/master.i I> update replication_synchro_quorum = 2
[092] 2022-11-15 00:01:29.678 [73463] main/112/applier/cluster@unix/:/tmp/t/092_replication/master.i I> RAFT: fencing paused
[092] 2022-11-15 00:01:29.678 [73463] main/112/applier/cluster@unix/:/tmp/t/092_replication/master.i I> final data received
[092] 2022-11-15 00:01:29.683 [73463] main C> got signal 15 - Terminated
[092] 2022-11-15 00:01:29.683 [73463] snapshot/101/main I> saving snapshot `/tmp/t/092_replication/replica_auth/00000000000000000007.snap.inprogress'
[092] 2022-11-15 00:01:29.683 [73463] main/117/iproto.shutdown I> tx_binary: stopped
[092] 2022-11-15 00:01:29.683 [73463] main I> Starting box_free()
[092] 2022-11-15 00:01:29.683 [73463] main I> Finished box_free()
[092] 2022-11-15 00:01:29.683 [73463] main I> Starting cbus_free()
[092] 2022-11-15 00:01:29.683 [73463] main I> Finished cbus_free()
[092] 2022-11-15 00:01:29.683 [73463] main I> Starting fiber_free()
[092] 2022-11-15 00:01:29.688 [73463] snapshot/101/main I> done
```
Apparently, for this particular case it helped with bypass the `at_exit()`.
And in the following case the snapshot is started right after the engine shutdown, which is inside the `box_free()`:
```
[070] 2022-11-15 00:34:38.380 [78819] main/112/applier/cluster@unix/:/tmp/t/070_replication/master.i I> final data received                          
[070] 2022-11-15 00:34:38.386 [78819] main C> got signal 15 - Terminated                                                                     
[070] 2022-11-15 00:34:38.386 [78819] main/117/iproto.shutdown I> tx_binary: stopped                                                                 
[070] 2022-11-15 00:34:38.386 [78819] main I> Starting box_free()             
[070] 2022-11-15 00:34:38.386 [78819] main I> Finished box_free()                   
[070] 2022-11-15 00:34:38.387 [78819] snapshot/101/main I> saving snapshot `/tmp/t/070_replication/replica_auth/00000000000000000007.snap.inprogress'
[070] 2022-11-15 00:34:38.387 [78819] main I> Starting cbus_free()            
[070] 2022-11-15 00:34:38.387 [78819] main I> Finished cbus_free()                                                                           
[070] 2022-11-15 00:34:38.387 [78819] main I> Starting fiber_free()                                                          
[070] 2022-11-15 00:34:38.392 [78819] snapshot/101/main I> done                                                                              
[070] [ fail ]                                                                
```
I will work on the changelog, but for the test I suppose it'll require some intervention with new ERRINJs to make the `tarantool_free()` and `cord_costart()` hold up for particular moment, orchestrating the situation above.
Can stabilization of the test be a good proof instead?

Reply:
> And in the following case the snapshot is started right after the engine shutdown, which is inside the `box_free()`

The checkpoint thread *logged* the "saving snapshot" message after `box_free`, but the checkpoint itself must have been started before `box_free` was called (it couldn't possibly be started after `box_free`, because the event loop was broken by that time). Still, I don't understand why `memtx_engine_shutdown` didn't cancel the checkpoint thread properly.

Reply:
The reason is quite simple: `box_free()` does engine shutdown only after it is being configured, while in test it happens _during_ the configuration - when it still not configured.

Reply:
Then fixing `cord_costart` doesn't seem to be enough, because the checkpoint thread may already be running and using some facilities of the main thread at the time when the signal is received. Maybe, we could run `box_free` if the initial box configuration completed (`engine_init` et al was called by `box_cfg`)?